### ISSUE TITLE
Runtime Instance shared by more than one App eg in different Tabs

### ISSUE THREAD

**Issue by pchainho** [reTHINK-project/dev-runtime-browser#66]
As agreed [here](https://github.com/reTHINK-project/specs/issues/15)

- [ ] windowSandbox implementation based on iFrame 

- [ ] Runtime deployment  in the Shared Worker and update interface communication between RuntimeUAStub and Shared Worker interface

- [ ] Update RuntimeUAStub to create, on runtimeUA request, iFrames or Web Workers, outside the runtime core and then deploy there components.

**Comment by dvilchez** [comment]
I made a proposal implementation on the window_sandbox branch, I think two new requirements are needed in the core:

1. A mechanism to pass the channel in the sandboxes creation.
2. An app sandbox per application

We can discuss this if you need more info

/cc @pchainho @vitormsilva

**Comment by dvilchez** [comment]
Communication between admin guis and runtime must use the postmessage interface since a direct communication is not possible anymore.

/cc @pchainho @AnaSLCaldeira

**Comment by pchainho** [comment]
> A mechanism to pass the channel in the sandboxes creation.

what do you mean by channel?

> An app sandbox per application

This one is aligned with https://github.com/reTHINK-project/dev-runtime-core/issues/136

**Comment by pchainho** [comment]
> Communication between admin guis and runtime must use the postmessage interface since a direct communication is not possible anymore.

[yes](https://github.com/reTHINK-project/dev-runtime-browser/issues/67)

**Comment by dvilchez** [comment]
The channel is the mean  through which messages between sandbox and its context are sended.

**Comment by pchainho** [comment]
![screen-shot-2016-12-12-13-14-48](https://cloud.githubusercontent.com/assets/3893553/21099720/b3b3da9e-c067-11e6-98c8-4bc5075c1967.png)

From the discussion in Slack we have concluded a Context is a Sandbox that is not the Core Sandbox eg App Sandbox, Web Workers and Window Sandbox.

Since:

* for each sandbox instance you have a pair of internal and external minibus instances that make the bridge between components deployed in the sandbox and all other components outside the sandbox though [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel)
* The external minibus is identified by the sandbox URL
* There is only one to one relationship between external and internal minibus, there only one MessageChannel per sandbox instance

I would say, the Sandbox URL would be enough. 

To be further clarified.

**Comment by pchainho** [comment]
pending on https://github.com/reTHINK-project/dev-runtime-core/issues/136 and https://github.com/reTHINK-project/dev-runtime-browser/issues/67.

**Comment by dvilchez** [comment]
> From the discussion in Slack we have concluded a Context is a Sandbox that is not the Core Sandbox eg App Sandbox, Web Workers and Window Sandbox.

No, in this design, **a sandbox** object is the **real sandbox facade**:

- a real sandbox is  an iframe, web worker or shared worker
- a sandbox is only an object that extends minibus and translates messages to and from the bus
- a context is the code loaded into a real sandbox

> for each sandbox instance you have a pair of internal and external minibus instances that make the bridge between components deployed in the sandbox and all other components outside the sandbox though MessageChannel

To be more clear, from the point of view of a hyperty sending a message:

The external minibus instance receives the message from the hyperty hosted in it and delegates this message to internal minibus through a MessageChannel. The internal sandbox receives the message from the MessageChannel and delegates it to the MessageBus. The MessageBus resolve the destination and delivers the messages to the right internal minibus.

Runtime core creates the internal minibuses and **request to the runtimeuastub** to create the external minibus for what we have to indicate the channel.

**Comment by vitormsilva** [comment]
To try resolve the problem, we have, with how to create a new app sandbox.

We propose the following:

1. runtime-browser create a rethink iframe on domain `rethink.com`
2. the rethink iframe, should create a shared worker
with the runtimeUA (beacuse the shared worker only can send messages to the same domain);
3. the shared worker, on the event "onconnect" (when a new app has connected) 
should invoque a new method from runtimeUA, `loadAppSandbox`.
4. this `loadAppSandbox` will receive one argument, the App URL address
5. this `loadAppSandbox` should create an AppSandbox, 
with a new address, like: `runtime://<runtime-domain>/<runtime-id>/<App URL without scheme>`
6. should return a promise with the result if fulfilled `runtime://<runtime-domain>/<runtime-id>/<App URL without scheme>` or reject.

Only one note, not important right now, but here: https://github.com/reTHINK-project/dev-runtime-browser/blob/window_sandbox/src/RuntimeUAStub.js#L71-L72
With this code, every time we did rethink.install.. a new shared worker will be created, because the blob will be always different.

**Comment by dvilchez** [comment]
What is this new appsandbox address for? As far as I know the sandboxes don't receive any mesasages from the runtime, the hyperties and protostubs do.

I can use the port received on the event to bind the port to the app sandbox, but as I told you this is not the only sandbox outside the shared worker, the runtime could create too windows sandboxes.

**Comment by vitormsilva** [comment]
> What is this new appsandbox address for?

The app sandbox address is for distinguishing each new app which could exist.


For each new sandbox, the responsibility is from **runtime-browser**, based on the capabilities previously obtained.

When the requireHyperty or requireStub is called, the installation process, in the *runtime-core*, will call the method `createSandbox(capabilities)` and the decision, if we shall create a **window sandbox** or **web worker**, is in this `createSandbox`.

something like this: https://github.com/reTHINK-project/dev-hyperty-toolkit/blob/dev-p2p/resources/factories/runtimeFactory.js#L18
Note: this is only a proposal..

**Comment by dvilchez** [comment]
>The app sandbox address is for distinguishing each new app which could exist.

how? could you provide a example message addressed to a hyperty inside an app sandbox?

>For each new sandbox, the responsibility is from runtime-browser, based on the capabilities previously obtained.

This has nothing to do with this discussion.

>When the requireHyperty or requireStub is called, the installation process, in the runtime-core, will call the method createSandbox(capabilities) and the decision, if we shall create a window sandbox or web worker, is in this createSandbox.

Runtime is still throwing the sandbox creation, and as I said before, I need a way to talk with the outside world.

>something like this: https://github.com/reTHINK-project/dev-hyperty-toolkit/blob/dev-p2p/resources/factories/runtimeFactory.js#L18

again, this has nothing to do with this discussion.

**Comment by pchainho** [comment]
@dvilchez can you comment here, pls, what was agreed in Vigo?

**Comment by dvilchez** [comment]
ok

@vitormsilva and I agreed on the runtime-core needs support the creation of multiple application sandboxes. However an important part is still missing, the implementation. We have to research a little deeper how shared worker works, we tried to do it but we couldn't manage to load a hyperty from the shared worker due to the window open popup call in the identity module.

**Comment by dvilchez** [comment]
Update

- [ ] Add loadApp(sandbox, appURL) to runtime and remove the code to create the application sandbox

- [ ] Create a new administration sandbox

- [ ] Call new loadApp method with application sandbox and with administration sandbox

- [ ] Load the window sandbox in any application context

Anything missed? @pchainho @vitormsilva

**Comment by pchainho** [comment]
seems ok

**Comment by pchainho** [comment]
who is creating the window sandox? the App?
